experimental/parser/parse_type.go

```diff
@@ -0,0 +1,196 @@
+		// keywords.
+		//
+		// This case applies to many other keywords. See cannotStartType.
+		if !isList && len(mods) == 0 &&
+			cannotStartType(ident.Text()) &&
+			!canStartPath(c.Peek()) {
+			kw, path := tyPath.Split(1)
+			if !path.Nil() {
+				return ast.TypePath{Path: kw}.AsAny(), path
+			}
+		}
+
+		// Check if ident is a modifier, and if so, peel it off.
+		//
+		// We need to be careful to only peel off `stream` inside of a method
+		// type. If the entire path is a single identifier, we always peel it
+		// off, since code that follows handles turning it back into a path
+		// based on what comes after it.
+		var isMod bool
+		_, rest := tyPath.Split(1)
+		switch ast.TypePrefixByName(ident.Name()) {
+		case ast.TypePrefixOptional, ast.TypePrefixRepeated, ast.TypePrefixRequired:
+			isMod = !isList || rest.Nil()
+		case ast.TypePrefixStream:
+			isMod = isInMethod || rest.Nil()
+		}
+
+		if isMod {
+			mods = append(mods, ident)
+			tyPath = rest
+		}
+	}
+
+	if tyPath.Nil() {
+		if len(mods) == 0 {
+			return ast.TypeAny{}, ast.Path{}
+		}
+
+		// Pop the last mod and make that into the type path. This makes
+		// `optional optional` work as a type.
+		last := mods[len(mods)-1]
+		tyPath = astx.NewPath(p.Context, last, last)
+		mods = mods[:len(mods)-1]
+	}
+
+	ty := ast.TypePath{Path: tyPath}.AsAny()
+
+	// Next, look for some angle brackets. We need to do this before draining
+	// mods, because angle brackets bind more tightly than modifiers.
+	if angles := c.Peek(); angles.Text() == "<" && !angles.IsLeaf() {
+		c.Pop() // Consume the angle brackets.
+		generic := p.NewTypeGeneric(ast.TypeGenericArgs{
+			Path:          tyPath,
+			AngleBrackets: angles,
+		})
+
+		commas := commas(p, angles.Children(), true, func(c *token.Cursor) (ast.TypeAny, bool) {
+			ty, _ := parseType(p, c, taxa.TypeParams.In(), false)
+			return ty, !ty.Nil()
+		})
+
+		commas(func(ty ast.TypeAny, comma token.Token) bool {
+			generic.Args().AppendComma(ty, comma)
+			return true
+		})
+
+		ty = generic.AsAny()
+	}
+
+	// Now, check for a path that follows all this. If there isn't a path, and
+	// ty is (still) a TypePath, and there is still at least one modifier, we
+	// interpret the last modifier as the type and the current path type as the
+	// path after the type.
+	var path ast.Path
+	if pathAfter {
+		next := c.Peek()
+		if canStartPath(next) {
+			path = parsePath(p, c)
+		} else if !isList && ty.Kind() == ast.TypeKindPath && len(mods) > 0 {
+			path = tyPath
+
+			// Pop the last mod and make that into the type. This makes
+			// `optional optional = 1` work as a proto3 field.
+			last := mods[len(mods)-1]
+			tyPath = astx.NewPath(p.Context, last, last)
+			mods = mods[:len(mods)-1]
+			ty = ast.TypePath{Path: tyPath}.AsAny()
+		}
+	}
+
+	// Finally, apply any remaining modifiers (in reverse order) to ty.
+	for i := len(mods) - 1; i >= 0; i-- {
+		ty = p.NewTypePrefixed(ast.TypePrefixedArgs{
+			Prefix: mods[i],
+			Type:   ty,
+		}).AsAny()
+	}
```

Comment:
Logically, I think this belongs above: after we've process any generic type arguments but before we look for an optional trailing path.

Reply:
It can't, and the comment in the previous stanza explains why. If we parse `optional optional`, and we want to return a path, we return `optional, optional`: a path type and a path. But if we don't want a path, we would return `optional optional, nil`: a prefixed type and a nil path.